### DISCUSSION TITLE
external files not loaded

### DISCUSSION THREAD

**Discussion by dannygueta** [fission/fission#2817] [Q&A]
Hi,

i'm trying to setup my first fission project, migrating a NodeJS AWS Lambda to Fission, i've built a package (zip) containing the following files

/main.js
/src/graphql/datamodel.graphql

the function itself runs but it fails to load /src/graphql/datamodel.graphql even though the file is clearly in the zip.

"level":"error","time":1688651708071,"pid":18,"hostname":"poolmgr-nodejs-default-25602273-66d7bf4d8f-l8wbv","version":"1.0.2","functionName":"prepareFirstBill","message":"user code load error: Error: ENOENT: no such file or directory, open './src/graphql/datamodel.graphql'"

i've downloaded the package using 'fission package getdeploy' and i can clearly see the zip file containing all of my source code yet during runtime it says the file is missing...

any help / lead ?

**Comment by sohaibaumer** [comment]
It appears that you're encountering an issue when trying to access the 'datamodel.graphql' file within your Fission function. The error message indicates that the file is not found, even though it exists in the package.
- You can double-check the file path you're using to access 'datamodel.graphql' within your Fission function. Ensure that the path is correct and properly reflects the location of the file within the package. Note that the current working directory might differ when running inside Fission.
- Verify the structure of your package and confirm that the 'datamodel.graphql' file is placed in the appropriate directory relative to the main entry file ('main.js'). Make sure the package structure aligns with the expected directory structure for your Fission function.
- Check the file permissions within the package. Ensure that the 'datamodel.graphql' file has the necessary read permissions, allowing it to be accessed by the Fission function during runtime.

**Reply by dannygueta** [reply]
Hi,

locally the code runs perfectly and finds the files, we've added some logs as we are trying to figure out whats going on.

we've added the following code in an attempt to try and debug whats going on:

```nodejs
console.log('dirname', __dirname);
console.log('env', process.env);

const path = require('path');
const fs = require('fs');

console.log('. =', path.resolve('.'));
console.log('__dirname = ', path.resolve(__dirname));
fs.readdirSync('.').forEach(file => {
  console.log(file);
});
```

the above code printed the following logs:
```
. = /usr/src/app
__dirname =  /usr/src/app/src/billCalculation/newBilling/firstBill
node_modules
package-lock.json
package.json
server.js
```

we've also done some digging into the container itself to see how the code is loaded, we can see its fetched and stored under /userfunc/ and the entire code is there.

```
/usr/src/app # ls -l /userfunc/deployarchive/
main.js      main.js.map  src/
/usr/src/app # ls -l /userfunc/deployarchive/src/graphql/
datamodel.graphql  generated/         whitelist.graphql
```

our conclusion is the code is running under /usrc/src/app while the code itself alongside its external files are loaded into /userfunc/.
this results in an error while trying to load external files.

is this expected  when using fission ? how should we reference external files relative to our source code ? 

Thanks.

**Comment by sohaibaumer** [comment]
Directly reference the absolute path to 'datamodel.graphql' within '/userfunc/'.

const fs = require('fs');
const fileContents = fs.readFileSync('/userfunc/src/graphql/datamodel.graphql', 'utf8');
console.log(fileContents);
